src/Kernel/Process.class.st

```diff
@@ -213,73 +213,31 @@ Process >> debugWithTitle: title [
-	This methods allow to not be restricted only on Semaphore case."
-	suspendedContext := oldList handleProcessTerminationOfWaitingContext: suspendedContext.
-
-	"If terminating a process halfways through an unwind, try to complete that unwind block first;
-	if there are multiple such nested unwind blocks, try to complete the outer-most one; the inner
-	blocks will be completed in the process."
-	ctxt := suspendedContext.
-	[(ctxt := ctxt findNextUnwindContextUpTo: nil) isNil] whileFalse: 
-		"Contexts under evaluation have already set their complete (tempAt: 2) to true."
-		[ctxt unwindComplete ifNotNil: [outerMost := ctxt]].
-	outerMost ifNotNil: [
-		"This is the outer-most unwind context currently under evaluation;
-		let's find an inner context executing outerMost's argument block (tempAt: 1)"
-		(suspendedContext findContextSuchThat: [:ctx | 
-			ctx closure == outerMost unwindBlock]) ifNotNil: [:inner | 
-				"Let's finish the unfinished unwind context only (i.e. up to inner) and return here"
-				outerMost := (suspendedContext runUntilErrorOrReturnFrom: inner) first.
-				"Update the receiver's suspendedContext (the previous step reset its sender to nil);
-				return, if the execution stack reached its bottom (e.g. in case of non-local returns)."
-				(suspendedContext := outerMost sender) ifNil: [^self]]]. 
-
-	"Now all unwind blocks caught halfway through have been completed; 
-	let's execute the ones still pending. Note: #findNextUnwindContextUpTo: starts
-	searching from the receiver's sender but the receiver itself may be an unwind context."
-	ctxt := suspendedContext.
-	ctxt isUnwindContext ifFalse: [ctxt := ctxt findNextUnwindContextUpTo: nil].
-	[ctxt isNil] whileFalse: [
-		ctxt unwindComplete ifNil: [
-			ctxt unwindComplete: true.
-			unwindBlock := ctxt unwindBlock.
-			"Create a context for the unwind block and execute it on the unwind block's stack. 
-			Note: using #value instead of #runUntilErrorOrReturnFrom: would lead to executing 
-			the unwind on the wrong stack preventing the correct execution of non-local returns."
-			suspendedContext := unwindBlock asContextWithSender: ctxt.
-			suspendedContext runUntilErrorOrReturnFrom: suspendedContext].
-		ctxt := ctxt findNextUnwindContextUpTo: nil].
-
-	"Mark the context as terminated"
-	suspendedContext terminateTo: nil.
-	suspendedContext setSender: nil receiver: self method: (Process>>#endProcess) arguments: {}
+	It assumes that self is not the active process and the termination is requested from another process"	
+		
+	[ ] ensure: [ | oldList context |
+		oldList := self suspend.
+		self isTerminated ifTrue: [^self].
+		self isTerminating ifTrue: [ProcessAlreadyTerminating signal. ^self].
+		terminating := true.
+		context := oldList handleProcessTerminationOfWaitingContext: suspendedContext.
+		suspendedContext := [context unwindTo: nil. self endProcess] asContext.
+		self priority: (Processor activePriority + 1 min: Processor highestPriority); resume
```

Comment:
This new logic looks suspicious. If we want the synchronization between killer process and killing process it should be explicit - logic based on priorities  is fragile .

Reply:
About failing test: 
I did not look very deep but I think that when Zinc server terminates at the end of #withServerDo: the termination does not really complete. Thus the final release of the port is performed concurrently with the next test starting another server during #withServerDo:. Therefore we have the intermittent failures here.

Reply:
Hi Denis,
It's indeed a fair assumption the process invoking termination will wait until the termination is completed. I missed the scenario when the process being terminated gets on a conditional variable during unwind and lets the killer process proceed :)
Thanks a lot for noticing!